common/src/main/kotlin/de/hpi/cloud/common/entity/Wrapper.kt

```diff
@@ -0,0 +1,175 @@
+                ),
+                value = value
+            )
+        }
+
+        inline fun <reified E : Entity<E>> jsonSerializerFor(): JsonSerializer<E> =
+            JsonSerializer(E::class.jsonSerializer())
+    }
+
+    @Serializer(forClass = Wrapper::class)
+    class JsonSerializer<E : Entity<E>>(
+        private val entitySerializer: KSerializer<E>
+    ) : KSerializer<Wrapper<E>> {
+        override val descriptor = object : SerialClassDescImpl("Wrapper") {
+            init {
+                addElement("type")
+                addElement("version")
+                addElement("id")
+                addElement("metadata")
+                addElement("value")
+            }
+        }
+
+        override fun serialize(encoder: Encoder, obj: Wrapper<E>) {
+            encoder.beginStructure(descriptor).let {
+                it.encodeStringElement(descriptor, 0, obj.type)
+                it.encodeIntElement(descriptor, 1, obj.version)
+                it.encodeStringElement(descriptor, 2, obj.id.value)
+                it.encodeSerializableElement(descriptor, 3, Metadata.serializer(), obj.metadata)
+                it.encodeSerializableElement(descriptor, 4, entitySerializer, obj.value)
+            }
+        }
+
+        override fun deserialize(decoder: Decoder): Wrapper<E> {
+            val dec = decoder.beginStructure(descriptor)
+            var type: String? = null
+            var version: Int? = null
+            var id: Id<E>? = null
+            var metadata: Metadata? = null
+            var value: E? = null
+
+            loop@ while (true) {
+                when (val i = dec.decodeElementIndex(descriptor)) {
+                    CompositeDecoder.READ_DONE -> break@loop
+                    0 -> type = dec.decodeStringElement(descriptor, i)
+                    1 -> version = dec.decodeIntElement(descriptor, i)
+                    2 -> id = Id(dec.decodeStringElement(descriptor, i))
+                    3 -> metadata = dec.decodeSerializableElement(descriptor, i, Metadata.serializer())
+                    4 -> value = dec.decodeSerializableElement(descriptor, i, entitySerializer)
+                    else -> throw SerializationException("Unknown index $i")
+                }
+            }
+            dec.endStructure(descriptor)
+            return Wrapper(
+                type ?: throw MissingFieldException("type"),
+                version ?: throw MissingFieldException("version"),
+                id ?: throw MissingFieldException("id"),
+                metadata ?: throw MissingFieldException("metadata"),
+                value ?: throw MissingFieldException("value")
+            )
+        }
+    }
+
+    val documentId: String
+        get() = id.documentId(type)
+
+    // region Mutation
+    fun withValue(context: Context, newValue: E): Wrapper<E> {
+        return if (value == newValue) withKeepAlive(context)
+        else copy(
+            metadata = metadata.copy(
```

Comment:
utility method for pushing an event?

Reply:
These are to ensure events are always generated. Without these methods you might just call `wrapper.copy(value = …)` and forget to add respective `UpdateEvent`.

Maybe we should refactor this to a plain class (without `data`) so the copy-method can't be used by other modules?